gnat2goto/driver/tree_walk.adb

```diff
@@ -1390,33 +1441,120 @@ package body Tree_Walk is
+
+                  Type_Loopvar : constant Irep := Do_Type_Reference
+                    (Etype (Etype (Defining_Identifier (Spec))));
+
+                  Sym_Loopvar : constant Irep :=
+                    Make_Symbol_Expr
+                      (Source_Location => Sloc (Defining_Identifier (Spec)),
+                       I_Type          => Type_Loopvar,
+                       Identifier      => Loopvar_Name);
+
+                  Init : constant Irep := New_Irep (I_Code_Assign);
+                  Cond : Irep;
+                  Post : Irep;
+
+                  Bound_Low  : constant Irep :=
+                    Do_Expression (Low_Bound (Dsd));
+                  Bound_High : constant Irep :=
+                    Do_Expression (High_Bound (Dsd));
+
+               begin
+                  --  Loop var decl
+                  Append_Op (Loop_Wrapper, Make_Code_Decl
+                             (Symbol => Sym_Loopvar,
+                              Source_Location =>
+                                Sloc (Defining_Identifier (Spec))));
+
+                  --  TODO: needs generalization to support enums
+                  if Reverse_Present (Spec) then
+                     Set_Lhs (Init, Sym_Loopvar);
+                     Set_Rhs (Init, Bound_High);
+                     Cond := Make_Op_Geq
+                       (Rhs             => Bound_Low,
+                        Lhs             => Sym_Loopvar,
+                        Source_Location => Sloc (Spec),
+                        Overflow_Check  => False,
+                        I_Type          => Make_Bool_Type,
+                        Range_Check     => False);
+                     Post := Make_Increment
+                       (Sym_Loopvar, Etype (Low_Bound (Dsd)), -1);
+                  else
+                     Set_Lhs (Init, Sym_Loopvar);
+                     Set_Rhs (Init, Bound_Low);
+                     Cond := Make_Op_Leq
+                       (Rhs             => Bound_High,
+                        Lhs             => Sym_Loopvar,
+                        Source_Location => Sloc (Spec),
+                        Overflow_Check  => False,
+                        I_Type          => Make_Bool_Type,
+                        Range_Check     => False);
+                     Post := Make_Increment
+                       (Sym_Loopvar, Etype (Low_Bound (Dsd)), 1);
+                  end if;
+                  Set_Source_Location (Init, Sloc (Spec));
+                  Set_Source_Location (Post, Sloc (Spec));
+                  Loop_Irep := Do_For_Statement (Init, Cond, Post);
+               end;
+            else
+               pragma Assert
+                 (Present (Iterator_Specification (Iter_Scheme)));
+               raise Program_Error; -- TODO: implement loop iterators
+            end if;
+         end if;
       end if;
 
+      Set_Loop_Body (Loop_Irep, Body_Block);
+
+      Append_Op (Loop_Wrapper, Loop_Irep);
+      if not Has_Created_Identifier (N) then
```

Comment:
That's a bit of optimization, to avoid generating too many labels.

To my understanding, `Has_Created_Identifier` indicates that the user has not given a name to the loop, but GNAT has done it internally. This implies that there cannot be an `exit <identifier>` statement, and therefore no label is required (I am placing a label right after named loops, such that we can use a `goto` to realize the `exit <identifier>` statement).

If that is too much dependency on GNAT internals, then we could simply remove the surrounding if statement and always create a label.